drivers/usb/gadget/function/f_scm.c

```diff
@@ -589,7 +647,417 @@ static void __exit f_scm_exit(void)
+ * free'd after this functions returns.
+ *
+ */
+void scm_proxy_recv_ack(struct scm_packet *packet, void *inst)
+{
+	struct scm_packet_list_entry *entry;
+	struct scm_packet *new_packet;
+	struct scm_proxy_inst *proxy_inst;
+
+	new_packet = kmalloc(sizeof(*packet), GFP_ATOMIC);
+	if (!new_packet)
+		return;
+	memcpy(new_packet, packet, sizeof(*packet));
+
+	entry = kmalloc(sizeof(struct scm_packet_list_entry),
+			GFP_ATOMIC);
+	if (!entry) {
+		kfree(new_packet);
+		return;
+	}
+
+	proxy_inst = inst;
+	INIT_LIST_HEAD(&entry->list_handle);
+	entry->packet = new_packet;
+
+	spin_lock(&(proxy_inst->ack_list_lock));
+	list_add(&proxy_inst->ack_list, &entry->list_handle);
+	spin_unlock(&(proxy_inst->ack_list_lock));
+
+	wake_up(&proxy_inst->ack_wait_queue);
+}
+EXPORT_SYMBOL_GPL(scm_proxy_recv_ack);
+
+/**
+ * scm_proxy_init - Initializes an instance of the SCM proxy
+ *
+ * @usb_context The USB context to link with this proxy instance
+ *
+ * Initializes an instance of the SCM proxy to allow the SCM USB driver to talk
+ * to the SCM network driver. The pointer returned by this function must be
+ * passed to all other exported proxy functions as the "proxy context" to allow
+ * the proxy to know which instance the operation is being performed on.
+ *
+ * Returns: A pointer to the instance for this proxy.
+ *
+ */
+void *scm_proxy_init(void *usb_context)
+{
+	struct scm_proxy_inst *proxy_inst;
+
+	proxy_inst = kzalloc(sizeof(struct scm_proxy_inst), GFP_KERNEL);
+	if (!proxy_inst)
+		return;
+	mutex_init(&proxy_inst->scm_msg_id_mutex);
+	init_waitqueue_head(&proxy_inst->ack_wait_queue);
+	proxy_inst->usb_context = usb_context;
+	proxy_inst->scm_msg_id = 0;
+
+	spin_lock_init(&proxy_inst->ack_list_lock);
+	INIT_LIST_HEAD(&proxy_inst->ack_list);
+
+	/* Start up the Xaptum SCM socket module */
+	xaprc00x_register(proxy_inst);
+
+	return proxy_inst;
+}
+EXPORT_SYMBOL_GPL(scm_proxy_init);
+
+/**
+ * scm_proxy_connect_socket - Connect an SCM socket
+ *
+ * @local_id The ID of the socket to close
+ * @addr The socket address
+ * @alen Address length in bytes
+ * @context The SCM proxy context
+ *
+ * Sends a command to the device to connect an SCM socket to a given address.
+ *
+ * Returns: 0 on success or returned SCM error code.
+ *
+ */
+int scm_proxy_connect_socket(int local_id, struct sockaddr *addr, int alen,
+	void *context)
+{
+	struct scm_packet *packet = kzalloc(sizeof(struct scm_packet),
+		GFP_KERNEL);
+	int ret;
+	struct scm_payload_ack ack;
+	struct scm_proxy_inst *proxy_inst;
+
+	proxy_inst = context;
+
+	packet->hdr.opcode = SCM_OP_CONNECT;
+	packet->hdr.msg_id = scm_proxy_get_msg_id(context);
+
+	if (addr->sa_family == AF_INET)
+		scm_proxy_assign_ip4(packet, addr);
+	else if (addr->sa_family == AF_INET6)
+		scm_proxy_assign_ip6(packet, addr);
+
+	scm_send_msg(packet,
+		sizeof(struct scm_packet_hdr) + packet->hdr.payload_len,
+		proxy_inst->usb_context);
+
+	ack = scm_proxy_wait_ack(packet->hdr.msg_id, context);
```

Comment:
If I'm reading this correctly, this function only supports blocking `connect`.  

For `enftun`, we need to support non-blocking `read` and `write`, so it probably makes sense to support non-blocking `connect` as well.

Ideally the USB stuff (`f_scm`) would be entirely non-blocking (aka asynchronous).  The socket layer (the kernel socket support or the `/net/scm/...` code) would emulate blocking behavior on top of that if needed.  I believe that's how the "real" drivers  (TCP, UDP, etc.) work and how our TCP proxy prototype worked.

Perhaps the USB code (`f_scm`) should store some sort of callback to the SCM code.  Then when the ACK is received, it can notify the socket layer?